pkg/process/process.go

```diff
@@ -0,0 +1,234 @@
+// Copyright 2022-2023 The Parca Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package process
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/prometheus/procfs"
+
+	"github.com/parca-dev/parca-agent/pkg/cgroup"
+)
+
+// key is a unique identifier for a process.
+// TODO(kakkoyun): This is not unique enough. We will need to add more fields.
+// For example, we can use the process start time to make sure we are not.
+// However, reading the stats file is expensive. For now, we will rely on the PID.
+// I will keep the struct to make it easier to add more fields in the future.
+type key struct {
+	pid int
+}
+
+type Tree struct {
+	tree        map[key]*process
+	mtx         *sync.RWMutex
+	disappeared map[key]struct{}
+	reused      map[key]*process
+}
+
+type process struct {
+	procfs.Proc
+
+	tree *Tree
+
+	parent int
+	// name of the field is the same as the one in kernel struct.
+	starttime uint64
+}
+
+func (p *process) ancestors() []*procfs.Proc {
+	p.tree.mtx.RLock()
+	defer p.tree.mtx.RUnlock()
+
+	ancestors := []*procfs.Proc{}
+	proc := p
+	for {
+		if proc.parent == 0 {
+			break
+		}
+		k := key{pid: proc.parent}
+		parent, ok := p.tree.tree[k]
+		if !ok {
+			break
+		}
+		ancestors = append(ancestors, &parent.Proc)
+		proc = parent
+	}
+	return ancestors
+}
+
+// NewTree returns a new process tree with current state of all the processes on the system.
+func NewTree(ctx context.Context, resetDuration time.Duration) *Tree {
+	t := &Tree{
+		tree:        make(map[key]*process),
+		disappeared: make(map[key]struct{}),
+		mtx:         &sync.RWMutex{},
+	}
+	go func() {
```

Comment:
If we do this the caller has no control over the lifetime of this goroutine. Let's move this to a `.Run(ctx)` function which the caller needs to call, and the caller decides whether to run it in a goroutine. I think this should probably even be a top-level goroutine managed by `run.Group`.

Reply:
It actually uses the same context and terminates the goroutine loop.
However, I like the idea of introducing a `Run(ctxt)` function. I'll do it.